ckb_smt.h

```diff
@@ -236,43 +247,96 @@ int smt_calculate_root(uint8_t *buffer, const smt_state_t *pairs,
           return ERROR_INVALID_PROOF;
         }
-        uint8_t height = proof[proof_index++];
+        uint16_t *height_a_ptr = &stack_heights[stack_top - 2];
+
+        uint16_t height_a = stack_heights[stack_top - 2];
         uint8_t *key_a = stack_keys[stack_top - 2];
         uint8_t *value_a = stack_values[stack_top - 2];
-        uint8_t *key_b = stack_keys[stack_top - 1];
+
+        uint16_t height_b = stack_heights[stack_top - 1];
         uint8_t *value_b = stack_values[stack_top - 1];
         stack_top -= 2;
-        int a_set = _smt_get_bit(key_a, height);
-        int b_set = _smt_get_bit(key_b, height);
-        _smt_copy_bits(key_a, height);
-        _smt_copy_bits(key_b, height);
-        uint8_t sibling_key_a[32];
-        memcpy(sibling_key_a, key_a, 32);
-        if (!a_set) {
-          _smt_set_bit(sibling_key_a, height);
+        if (height_a != height_b) {
+          return ERROR_INVALID_PROOF;
         }
-        if (memcmp(sibling_key_a, key_b, 32) != 0 || (a_set == b_set)) {
-          return ERROR_INVALID_SIBLING;
+        if (height_a > 255) {
+          return ERROR_INVALID_PROOF;
         }
-        if (a_set) {
-          _smt_merge(value_b, value_a, value_a);
+        uint8_t parent_key[SMT_KEY_BYTES];
+        memcpy(parent_key, key_a, SMT_KEY_BYTES);
+        _smt_parent_path(parent_key, (uint8_t)height_a);
+
+        // push value
+        if (_smt_get_bit(key_a, height_a)) {
+          _smt_merge(height_a, parent_key, value_b, value_a, value_a);
         } else {
-          _smt_merge(value_a, value_b, value_a);
+          _smt_merge(height_a, parent_key, value_a, value_b, value_a);
         }
-        /* Top-of-stack key is already updated to parent_key_a */
+        // push key
+        _smt_parent_path(key_a, height_a);
+        // push height
+        *height_a_ptr = height_a + 1;
         stack_top++;
       } break;
+      case 0x4F: {
+        if (stack_top < 1) {
+          return ERROR_INVALID_STACK;
+        }
+        if (proof_index >= proof_length) {
+          return ERROR_INVALID_PROOF;
+        }
+        uint16_t n = proof[proof_index];
+        proof_index++;
+        uint16_t zero_count = 0;
+        if (n == 0) {
+          zero_count = 256;
+        } else {
+          zero_count = n;
+        }
+        uint16_t *base_height_ptr = &stack_heights[stack_top - 1];
+        uint16_t base_height = stack_heights[stack_top - 1];
+        uint8_t *key = stack_keys[stack_top - 1];
+        uint8_t *value = stack_values[stack_top - 1];
+        if (base_height > 255) {
+          return ERROR_INVALID_PROOF;
+        }
+        uint8_t parent_key[SMT_KEY_BYTES];
+        memcpy(parent_key, key, SMT_KEY_BYTES);
+        uint16_t height_u16 = base_height;
+        for (uint16_t idx = 0; idx < zero_count; idx++) {
+          height_u16 = base_height + idx;
+          if (height_u16 > 255) {
+            return ERROR_INVALID_PROOF;
+          }
+          memcpy(parent_key, key, SMT_KEY_BYTES);
```

Comment:
No need to copy the `key` again, just use old `parent_key`

Reply:
It's tricky and not necessary to do it. It's not very easy to maintain.

Reply:
fixed. Add  more comments